include/nbl/ui/CWindowWin32.h

```diff
@@ -39,97 +41,150 @@ class CWindowWin32 final : public IWindowWin32
-			style |= WS_CLIPCHILDREN;
-			style |= WS_CLIPSIBLINGS;
-		}
-		if (isMinimized())
-			style |= WS_MINIMIZE;
-		if (isMaximized())
-			style |= WS_MAXIMIZE;
-		// TODO:
-		// if (hasMouseCaptured())
-		// if (hasInputFocus())
-		// if (hasMouseFocus())
-		// if (isAlwaysOnTop())
-
-		AdjustWindowRect(&clientSize, style, FALSE);
-
-		const int32_t realWidth = clientSize.right - clientSize.left;
-		const int32_t realHeight = clientSize.bottom - clientSize.top;
-
-		int32_t windowLeft = (GetSystemMetrics(SM_CXSCREEN) - realWidth) / 2;
-		int32_t windowTop = (GetSystemMetrics(SM_CYSCREEN) - realHeight) / 2;
-
-		if (windowLeft < 0)
-			windowLeft = 0;
-		if (windowTop < 0)
-			windowTop = 0;	// make sure window menus are in screen on creation
-
-		if (isFullscreen())
-		{
-			windowLeft = 0;
-			windowTop = 0;
-		}
 
-		// create window
+			// TODO:
+			// if (hasMouseCaptured())
+			// if (hasInputFocus())
+			// if (hasMouseFocus())
 
-		m_native = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
-			realWidth, realHeight, NULL, NULL, hinstance, NULL);
-		m_width = realWidth;
-		m_height = realHeight;
+			AdjustWindowRect(&clientSize, style, FALSE);
 
-		if (!isHidden())
-			ShowWindow(m_native, SW_SHOWNORMAL);
-		UpdateWindow(m_native);
+			const int32_t realWidth = clientSize.right - clientSize.left;
+			const int32_t realHeight = clientSize.bottom - clientSize.top;
 
-		// fix ugly ATI driver bugs. Thanks to ariaci
-		// TODO still needed?
-		MoveWindow(m_native, windowLeft, windowTop, realWidth, realHeight, TRUE);
-    }
+			int32_t windowLeft = (GetSystemMetrics(SM_CXSCREEN) - realWidth) / 2;
+			int32_t windowTop = (GetSystemMetrics(SM_CYSCREEN) - realHeight) / 2;
+
+			if (windowLeft < 0)
+				windowLeft = 0;
+			if (windowTop < 0)
+				windowTop = 0;	// make sure window menus are in screen on creation
+
+			if (_flags & ECF_FULLSCREEN)
+			{
+				windowLeft = 0;
+				windowTop = 0;
+			}
+
+			// create window
+
+			 nativeWindow = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
+				realWidth, realHeight, NULL, NULL, hinstance, NULL);
+			if ((_flags & ECF_HIDDEN) == 0)
+				ShowWindow(nativeWindow, SW_SHOWNORMAL);
+			UpdateWindow(nativeWindow);
+
+			// fix ugly ATI driver bugs. Thanks to ariaci
+			// TODO still needed?
+			MoveWindow(nativeWindow, windowLeft, windowTop, realWidth, realHeight, TRUE);
+		}
 
-    native_handle_t m_native;
+		native_handle_t getNativeWindow() const
+		{
+			return nativeWindow;
+		}
+	private:
+		void init()
+		{
+			// All the stuff was already initialized in the ctor
+		}
+		void exit()
+		{
+			DestroyWindow(nativeWindow);
+		}
+		void work(base_t::lock_t& lock)
+		{
+			// @criss not sure why PeekMessage is better than GetMessage, since we'll be spining all the time here
+			// Is a spinlock a better solution than blocking a thread?
+			keepRunning = GetMessage(&message, nativeWindow, 0, 0);
+			TranslateMessage(&message);
+			DispatchMessage(&message);
```

Comment:
But ofc you're right spinlock is bad, 

https://stackoverflow.com/questions/10866311/getmessage-with-a-timeout

Use SetTimer to emulate a timeout on GetMessage. 50ms should be reasonable since window creation/destruction is gonna be extremely rare request
@devshgraphicsprogramming ?

Reply:
> Use SetTimer to emulate a timeout on GetMessage. 50ms should be reasonable since window creation/destruction is gonna be extremely rare request

We can presume that user will want to run at 60 to 120 Hz, so a timeout of 8ms sounds practical.

P.S. Instead of setting up and killing a timer (who knows what that allocates), maybe just call a `PostMessage` with a custom/dummy message type from the requesting thread to wake it up from the `GetMessage` sleep.

Reply:
@devshgraphicsprogramming apparently 8ms is not possible since the minimum timeout (`USER_TIMER_MINIMUM`) is 10ms.

Reply:
can you `PostMessage` from the request submitting thread instead?

Reply:
thats a neat idea, nice. But... are you sure you can PostMessage w/o a window + this message will be catchable by this other thread?

Reply:
you can just post the message to any odd window, and your GetMessage will be woken to handle it

AFAIK PostMessage can be called from a different thread.

Reply:
ok, @sadiuk make sure its right and go with this approach

Reply:
btw theres a `PostMessage` and a `PostMessageThread`, I havent spent enough time looking into their documentation to know whats the difference.